### ISSUE TITLE
cursor warping incorrectly when constraint deactivated

### ISSUE THREAD

**Issue by s-viour** [hyprwm/Hyprland#5055]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
❯ hyprctl version
Hyprland, built from branch main at commit 26cd1bf949d0460b754fe3b1ae7f06bd4b2535ee dirty (input: fix minor default cursor reset conditions).
Date: Sat Mar 9 12:12:55 2024
Tag: v0.36.0-89-g26cd1bf9

flags: (if any)
```

</details>

### Bug or Regression?

Regression

### Description

I was previously on version 0.36.0 (before rewritten pointer constraints https://github.com/hyprwm/Hyprland/pull/4889) and everything worked fine. I tried updating today to the latest commit, and noticed that my cursor in FFXIV would warp to the previous position I started dragging the camera around from whenever I clicked on the screen. I assume this is because FFXIV attempts to constrain the pointer pretty much anytime you click? It doesn't happen when dragging in-game windows around, only when you click somewhere on the screen where you'd normally be able to drag the camera around.

I don't know a lot about wlroots programming or WMs in general, although I did some digging and found that the bug is almost definitely part of `CConstraint::deactivate`, as just commenting out the `warpCursorTo` call fixes the issue completely for me:
```diff
diff --git a/src/desktop/Constraint.cpp b/src/desktop/Constraint.cpp
index ea89037f..bc06c124 100644
--- a/src/desktop/Constraint.cpp
+++ b/src/desktop/Constraint.cpp
@@ -94,7 +94,7 @@ void CConstraint::deactivate() {
 
     wlr_pointer_constraint_v1_send_deactivated(m_pConstraint);
     m_bActive = false;
-    g_pCompositor->warpCursorTo(logicPositionHint(), true);
+    //g_pCompositor->warpCursorTo(logicPositionHint(), true);
 
     if (m_pConstraint->lifetime == ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT)
         m_bDead = true;
```

I'd submit a PR but I dunno enough to fix this without deleting the call altogether \:p. 

### How to reproduce

Open FFXIV -> click and drag on title screen to "save" your mouse position -> click somewhere else on screen and watch your mouse teleport to the position you started dragging from

**Comment by vaxerski** [comment]
probably should be only when locked?

**Comment by s-viour** [comment]
Looks like that does it! Following patch fixes the issue:
```diff
diff --git a/src/desktop/Constraint.cpp b/src/desktop/Constraint.cpp
index ea89037f..662f0990 100644
--- a/src/desktop/Constraint.cpp
+++ b/src/desktop/Constraint.cpp
@@ -94,7 +94,9 @@ void CConstraint::deactivate() {
 
     wlr_pointer_constraint_v1_send_deactivated(m_pConstraint);
     m_bActive = false;
-    g_pCompositor->warpCursorTo(logicPositionHint(), true);
+
+    if (isLocked())
+        g_pCompositor->warpCursorTo(logicPositionHint(), true);
 
     if (m_pConstraint->lifetime == ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT)
         m_bDead = true;

```

Want me to toss up a PR?

**Comment by vaxerski** [comment]
feel free

**Comment by qweered** [comment]
fixed #5055

**Comment by qweered** [comment]
Fixed in #5056
